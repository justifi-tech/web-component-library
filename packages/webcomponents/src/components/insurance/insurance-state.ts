import { createStore } from '@stencil/store';

const { state: insuranceValues, on: insuranceValuesOn } = createStore<any>({});

const { state: insuranceErrors } = createStore<any>({});

const validateInsuranceValues = () => {
  let invalid = false;

  Object.keys(insuranceValues).forEach((key) => {
    if (insuranceValues[key]) {
      insuranceErrors[key] = false;
    } else {
      insuranceErrors[key] = true;
      invalid = true;
    }
  });

  return { isValid: !invalid };
};

export { insuranceValues, insuranceValuesOn, insuranceErrors, validateInsuranceValues };
