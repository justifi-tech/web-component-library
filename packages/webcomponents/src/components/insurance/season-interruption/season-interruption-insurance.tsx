import { Component, Event, EventEmitter, Method, Prop, State, h } from '@stencil/core';
import { InsuranceService } from '../../../api/services/insurance.service';
import { makeGetQuote, makeToggleCoverage } from '../../../actions/insurance/insurance-actions';
import { ComponentErrorCodes, ComponentErrorSeverity } from '../../../api/ComponentError';
import JustifiAnalytics from '../../../api/Analytics';
import { checkPkgVersion } from '../../../utils/check-pkg-version';
import { ComponentErrorEvent } from '../../../api/ComponentEvents';
import {
  text,
  textDanger,
  insuranceForm,
  insuranceRadioGroup,
  insuranceHeaderSection,
  insuranceDescriptionSection,
  insuranceFormSection,
  insuranceLegalDisclaimerSection,
  insuranceLoading,
  inputRadio,
  inputRadioChecked,
  inputRadioCheckedFocused,
  inputRadioFocused,
  inputRadioInvalid,
  label,
} from '../../../styles/parts';

@Component({
  tag: 'justifi-season-interruption-insurance',
  shadow: true,
})
export class SeasonInterruptionInsurance {
  @Prop() authToken: string;
  @Prop() checkoutId: string;
  @Prop() primaryIdentityFirstName: string;
  @Prop() primaryIdentityLastName: string;
  @Prop() primaryIdentityState: string;
  @Prop() primaryIdentityPostalCode: string;
  @Prop() primaryIdentityCountry: string;
  @Prop() primaryIdentityEmailAddress: string;
  @Prop() policyAttributesInsurableAmount: number;
  @Prop() policyAttributesStartDate?: string;
  @Prop() policyAttributesEndDate?: string;
  @Prop() coveredIdentityFirstName?: string;
  @Prop() coveredIdentityLastName?: string;

  @State() getQuote: Function;
  @State() toggleCoverage: Function;
  @State() quote: any;

  @Event({ eventName: 'error-event' }) errorEvent: EventEmitter<ComponentErrorEvent>;

  private seasonInterruptionCoreRef?: HTMLJustifiSeasonInterruptionInsuranceCoreElement;

  analytics: JustifiAnalytics;

  @Method()
  async validate(): Promise<{ isValid: boolean }> {
    return this.seasonInterruptionCoreRef.validate();
  }

  componentWillLoad() {
    checkPkgVersion();
    this.analytics = new JustifiAnalytics(this);
    this.initializeServiceMethods();
  }

  disconnectedCallback() {
    this.analytics?.cleanup();
  }

  private initializeServiceMethods() {
    if (!this.authToken) {
      this.errorEvent.emit({
        errorCode: ComponentErrorCodes.MISSING_PROPS,
        message: 'Missing authToken',
        severity: ComponentErrorSeverity.ERROR,
      });
      return;
    }

    this.getQuote = makeGetQuote({
      authToken: this.authToken,
      service: new InsuranceService(),
    });

    this.toggleCoverage = makeToggleCoverage({
      authToken: this.authToken,
      service: new InsuranceService(),
    });
  }

  // Join the parts, split on spaces to get individual part names,
  // filter out any extra spaces to be safe, then rejoin properly
  private usedPartsString = [
    text,
    textDanger,
    insuranceForm,
    insuranceRadioGroup,
    insuranceHeaderSection,
    insuranceDescriptionSection,
    insuranceFormSection,
    insuranceLegalDisclaimerSection,
    insuranceLoading,
    inputRadio,
    inputRadioChecked,
    inputRadioCheckedFocused,
    inputRadioFocused,
    inputRadioInvalid,
    label,
  ].join(' ').split(' ').filter(part => part.length > 0).join(', ');

  render() {
    return (
      <justifi-season-interruption-insurance-core
        ref={(el) => (this.seasonInterruptionCoreRef = el)}
        checkout-id={this.checkoutId}
        getQuote={this.getQuote}
        toggleCoverage={this.toggleCoverage}
        primary-identity-first-name={this.primaryIdentityFirstName}
        primary-identity-last-name={this.primaryIdentityLastName}
        primary-identity-state={this.primaryIdentityState}
        primary-identity-postal-code={this.primaryIdentityPostalCode}
        primary-identity-country={this.primaryIdentityCountry}
        primary-identity-email-address={this.primaryIdentityEmailAddress}
        policy-attributes-insurable-amount={this.policyAttributesInsurableAmount}
        policy-attributes-start-date={this.policyAttributesStartDate}
        policy-attributes-end-date={this.policyAttributesEndDate}
        covered-identity-first-name={this.coveredIdentityFirstName}
        covered-identity-last-name={this.coveredIdentityLastName}
        exportparts={this.usedPartsString}
      >
      </justifi-season-interruption-insurance-core>
    );
  }
}
