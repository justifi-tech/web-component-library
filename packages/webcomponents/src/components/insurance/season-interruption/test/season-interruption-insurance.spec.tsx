jest.mock('../../../../ui-components/styled-host/styled-host.css', () => '');

import { h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';
import { SeasonInterruptionInsurance } from '../season-interruption-insurance';
import { insuranceValuesStore, insuranceErrorsStore } from '../../insurance-state';

describe('justifi-season-interruption-insurance', () => {
  beforeEach(() => insuranceValuesStore.dispose());
  beforeEach(() => insuranceErrorsStore.dispose());

  it('should display loading state correctly', async () => {
    const page = await newSpecPage({
      components: [SeasonInterruptionInsurance],
      template: () => <justifi-season-interruption-insurance auth-token="123" />,
    });

    await page.waitForChanges();

    expect(page.root).toMatchSnapshot();
  });

  it('loads and sets the quote to state correctly', async () => {
    const page = await newSpecPage({
      components: [SeasonInterruptionInsurance],
      template: () => <justifi-season-interruption-insurance auth-token="123" />,
    });

    await page.waitForChanges();

    expect(page.root).toMatchSnapshot();
  });

  it('should emit error event when fetch fails', async () => {
    const errorSpy = jest.fn();

    const page = await newSpecPage({
      components: [SeasonInterruptionInsurance],
      template: () => <justifi-season-interruption-insurance auth-token="" onError-event={errorSpy} />,
    });

    await page.waitForChanges();
    await new Promise(resolve => setTimeout(resolve, 100)); // Wait for async operations

    expect(errorSpy).toHaveBeenCalledWith(
      expect.objectContaining({
        detail: {
          errorCode: 'missing-props',
          message: 'Missing authToken',
          severity: 'error',
        }
      })
    );
  });

  it('validates that a selection was made', async () => {
    const page = await newSpecPage({
      components: [SeasonInterruptionInsurance],
      template: () => <justifi-season-interruption-insurance auth-token="123" />,
    });

    await page.waitForChanges();

    const instance: any = page.rootInstance;
    const { isValid } = await instance.validate();

    await page.waitForChanges();

    expect(isValid).toBe(true); // The validation should pass when no selection is made initially
    expect(page.root).toMatchSnapshot();
  });
});
