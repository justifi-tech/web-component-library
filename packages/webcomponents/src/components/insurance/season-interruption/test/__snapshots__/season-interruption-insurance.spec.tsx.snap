// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`justifi-season-interruption-insurance renders loading 1`] = `
<justifi-season-interruption-insurance auth-token="123">
  <template shadowrootmode="open">
    <justifi-season-interruption-insurance-core checkout-id="" class="body" style="display: block;">
      <style></style>
      <div class="gap-3 row">
        <div class="col-12">
          <div aria-hidden="true" class="skeleton undefined" part="skeleton" style="width: 35%; height: 30px;"></div>
        </div>
        <div class="col-12">
          <div aria-hidden="true" class="skeleton undefined" part="skeleton" style="width: 70%;"></div>
        </div>
        <div class="col-12">
          <div aria-hidden="true" class="skeleton undefined" part="skeleton" style="width: 40%;"></div>
        </div>
        <div class="col-12">
          <div aria-hidden="true" class="skeleton undefined" part="skeleton" style="width: 40%;"></div>
        </div>
        <div class="col-12">
          <div aria-hidden="true" class="skeleton undefined" part="skeleton" style="width: 100%;"></div>
        </div>
        <div class="col-12">
          <div aria-hidden="true" class="skeleton undefined" part="skeleton" style="width: 80%;"></div>
        </div>
      </div>
    </justifi-season-interruption-insurance-core>
  </template>
</justifi-season-interruption-insurance>
`;
